### PR TITLE
[ws-manager-mk2] Protect environment secrets

### PR DESCRIPTION
**Pull request by Furisto**
## Description
Protect environment secrets

## Related Issue(s)
n.a.

## How to test
- Set environment variable in user preferences
- Create workspace
- Check that workspace CR contains reference to the secret
- Check that secret is deleted after workspace has started
- Check environment variables in workspace

## Release Notes

```release-note
None
```

## Build Options:

- [ ] /werft with-github-actions
      Experimental feature to run the build with GitHub Actions (and not in Werft).
- [ ] leeway-no-cache
      leeway-target=components:all
- [ ] /werft no-test
      Run Leeway with `--dont-test`

<details>
<summary>Publish Options</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer Options</summary>

- [ ] with-ee-license
- [ ] with-slow-database
- [ ] with-dedicated-emulation
- [ ] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

**Comment by Furisto** [comment]
/werft run with-preview with-large-vm with-wsman-mk2 recreate-vm=true

File: components/ws-manager-mk2/controllers/workspace_controller.go [hunk 1]
```diff
@@ -170,6 +171,8 @@ func (r *WorkspaceReconciler) actOnStatus(ctx context.Context, workspace *worksp
 				}
 			}
 
+			r.deleteWorkspaceSecrets(ctx, workspace)
```
Comment (WVerlaek) [inline components/ws-manager-mk2/controllers/workspace_controller.go:174]: lets check for an error, and not delete the workspace if there was an issue removing the secrets so we try again in the next reconcile tick

**Review by WVerlaek — COMMENTED** [review]
quick pass - will do some testing tomorrow

**Comment by Furisto** [comment]
> and like we chatted: could you add some tests for the creation + deletion of the secret pray

Done

File: components/ws-manager-mk2/controllers/workspace_controller.go (hunk 1 above)
Comment (Furisto) [reply components/ws-manager-mk2/controllers/workspace_controller.go:186]: Will add a retry for secret deletion in the token secret PR.

File: components/ws-manager-mk2/service/manager.go [hunk 2]
```diff
@@ -271,6 +279,67 @@ func (wsm *WorkspaceManagerServer) StartWorkspace(ctx context.Context, req *wsma
...
+	}
+
+	err := wsm.Client.Create(ctx, &secret)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return err
+	}
+
+	err = wait.PollWithContext(ctx, 100*time.Millisecond, 5*time.Second, func(c context.Context) (done bool, err error) {
+		var secret corev1.Secret
+		err = wsm.Client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: name}, &secret)
```
Comment (WVerlaek) [inline components/ws-manager-mk2/service/manager.go:324]:
```suggestion
		err = wsm.Client.Get(c, types.NamespacedName{Namespace: namespace, Name: name}, &secret)
```

File: components/ws-manager-mk2/service/manager.go [hunk 3]
```diff
@@ -271,6 +279,67 @@ func (wsm *WorkspaceManagerServer) StartWorkspace(ctx context.Context, req *wsma
...
+		},
+		StringData: data,
+	}
+
+	err := wsm.Client.Create(ctx, &secret)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return err
+	}
+
+	err = wait.PollWithContext(ctx, 100*time.Millisecond, 5*time.Second, func(c context.Context) (done bool, err error) {
```
Comment (WVerlaek) [inline components/ws-manager-mk2/service/manager.go:322]: why the poll actually? If the `Create` succeeded, do we need to check again?

**Review by WVerlaek — COMMENTED** [review]
lgtm otherwise

File: components/ws-manager-mk2/service/manager.go (hunk 3 above)
Comment (Furisto) [reply components/ws-manager-mk2/service/manager.go]: Actually not sure why I did that anymore :sweat_smile: . Removed it.

**Review by WVerlaek — APPROVED** [review]

**Pull request merged**